nixos/modules/services/networking/ssh/sshd.nix

```diff
@@ -539,204 +789,306 @@ in
-                      fi
-                      mkdir -p "$(dirname '${k.path}')"
-                      chmod 0755 "$(dirname '${k.path}')"
-                      ssh-keygen \
-                        -t "${k.type}" \
-                        ${lib.optionalString (k ? bits) "-b ${toString k.bits}"} \
-                        ${lib.optionalString (k ? rounds) "-a ${toString k.rounds}"} \
-                        ${lib.optionalString (k ? comment) "-C '${k.comment}'"} \
-                        ${lib.optionalString (k ? openSSHFormat && k.openSSHFormat) "-o"} \
-                        -f "${k.path}" \
-                        -N ""
-                  fi
-                '')}
-              '';
-
-            serviceConfig =
-              { ExecStart =
-                  (lib.optionalString cfg.startWhenNeeded "-") +
-                  "${cfg.package}/bin/sshd " + (lib.optionalString cfg.startWhenNeeded "-i ") +
-                  "-D " +  # don't detach into a daemon process
-                  "-f /etc/ssh/sshd_config";
-                KillMode = "process";
-              } // (if cfg.startWhenNeeded then {
-                StandardInput = "socket";
-                StandardError = "journal";
-              } else {
-                Restart = "always";
-                Type = "simple";
-              });
+        service = {
+          description = "SSH Daemon";
+          wantedBy = lib.optional (!cfg.startWhenNeeded) "multi-user.target";
+          after = [ "network.target" ];
+          stopIfChanged = false;
+          path = [
+            cfg.package
+            pkgs.gawk
+          ];
+          environment.LD_LIBRARY_PATH = nssModulesPath;
 
-          };
+          restartTriggers = lib.optionals (!cfg.startWhenNeeded) [
+            config.environment.etc."ssh/sshd_config".source
+          ];
+
+          preStart = ''
+            # Make sure we don't write to stdout, since in case of
+            # socket activation, it goes to the remote side (#19589).
+            exec >&2
+
+            ${lib.flip lib.concatMapStrings cfg.hostKeys (k: ''
+              if ! [ -s "${k.path}" ]; then
+                  if ! [ -h "${k.path}" ]; then
+                      rm -f "${k.path}"
+                  fi
+                  mkdir -p "$(dirname '${k.path}')"
+                  chmod 0755 "$(dirname '${k.path}')"
+                  ssh-keygen \
+                    -t "${k.type}" \
+                    ${lib.optionalString (k ? bits) "-b ${toString k.bits}"} \
+                    ${lib.optionalString (k ? rounds) "-a ${toString k.rounds}"} \
+                    ${lib.optionalString (k ? comment) "-C '${k.comment}'"} \
+                    ${lib.optionalString (k ? openSSHFormat && k.openSSHFormat) "-o"} \
+                    -f "${k.path}" \
+                    -N ""
+              fi
+            '')}
+          '';
+
+          serviceConfig =
+            {
+              CapabilityBoundingSet = [
```

Comment:
Why are you not using the capability groups? Also we shouldn't do one commit per capability. Probably most of them can be combined into one or two commits to not have a copy pasted repetitive history.

Reply:
1. This is a draft PR, please leave feedback until it is ready for review.
2. Because someone would complain if I put a bunch of them in the same commit.
3. Because reverting is infinitely easier if it's done one commit per capability
4. I'm not an expert on systemd or sshd, so I'm simply following the recommendations of `systemd-analyze security sshd`.
5. This is not "repeated history", it's one capability per commit, for ease of reverting.
6. Please explain what you mean by "capability groups". That term is so vague that I'm not going to go on a wild goose hunt trying to guess at what you meant.